src/catch2/internal/catch_compiler_capabilities.hpp

```diff
@@ -57,6 +57,11 @@
 
 #endif
 
+#if defined(__CUDACC__) && !defined(__clang__)
+#    define CATCH_INTERNAL_START_WARNINGS_SUPPRESSION _Pragma( "nv_diag_suppress 177" )
+#    define CATCH_INTERNAL_STOP_WARNINGS_SUPPRESSION  _Pragma( "nv_diag_default 177" )
```

Comment:
So, this would work, but wouldn't properly slot into how the macros are used across Catch2. If I am [reading the docs correctly](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#nv-diagnostic-pragmas), then the correct approach would be something like this


```suggestion
#    define CATCH_INTERNAL_START_WARNINGS_SUPPRESSION _Pragma( "nv_diagnostic push" )
#    define CATCH_INTERNAL_STOP_WARNINGS_SUPPRESSION  _Pragma( "nv_diagnostic pop" )
#    define CATCH_INTERNAL_SUPPRESS_UNUSED_VARIABLE_WARNINGS _Pragma ( "nv_diag_suppress 177" )
```

Reply:
Thanks for your comment! Unfortunately, your suggestion makes the warnings come back. The reason is that in the place where the "unused" variables are introduced we do not use `CATCH_INTERNAL_SUPPRESS_UNUSED_VARIABLE_WARNINGS`, but `CATCH_INTERNAL_SUPPRESS_GLOBALS_WARNINGS`. Apparently, other compilers than nvcc would emit a different warning here. I see to solutions to this:

1. Define `CATCH_INTERNAL_SUPPRESS_GLOBALS_WARNINGS` with the suppression of 177 instead of `CATCH_INTERNAL_SUPPRESS_UNUSED_VARIABLE_WARNINGS`.
2. Leave the `#define`s as you suggested and add `CATCH_INTERNAL_SUPPRESS_UNUSED_VARIABLE_WARNINGS` to the macros in `catch_test_registry.hpp` that introduce the variables nvcc complains about.

I updated my PR to implement option 2), since I think that is the cleaner approach. After all, nvcc's warning is about a variable that is "declared but never referenced" and nothing related to global variables. Let me know what you think!

Reply:
2. is the right solution, but it might be worth it to open up a bug upstream that a type with opaque constructor triggers unused variable warning. Usually compilers don't warn on this, because they assume that the opaque constructor contains side-effects that the user cares for.